File: src/main/digesthistory.ts
/**
 * Dlouhodobá čísla k přehledu.
 *
 * Třicet dní samo o sobě neřekne, jestli je to hodně nebo málo. „Sto třináct
 * objednávek" je dobrá zpráva v lednu a špatná v prosinci — a poznat se to dá
 * jedině proti tomu, co bylo dřív. Tenhle modul proto drží **měsíční souhrny
 * za celou historii ve feedu** a z nich počítá tři věci, které samotné okno
 * neumí:
 *
 *  1. **Zasazení do roku** — kolik z posledních dvanácti měsíců bylo slabších
 *     než těch posledních třicet dní.
 *  2. **Srovnání s loňskem** — stejné okno o rok zpátky. To je jediné
 *     srovnání, které nemate sezónou: prosinec proti prosinci, ne proti
 *     listopadu.
 *  3. **Sezóny** — které měsíce bývaly silné a kdy se na ně má začít chystat.
 *     Nic se nehádá dopředu podle kalendáře: index se počítá z **vlastních
 *     dat** e-shopu, takže Vánoce, svatby i letní útlum vyjdou samy, pokud
 *     v číslech doopravdy jsou.
 *
 * ## Proč se to ukládá
 *
 * Projít tisíce objednávek při každém otevření přehledu je zbytečné, když se
 * uzavřené měsíce už nezmění. Spočítané měsíce se proto drží v tabulce
 * `digest_months` a přepočítává se jen ten rozdělaný — a ten poslední, do
 * kterého ještě mohly dojít objednávky.
 */
import { getDb } from './db';
import { bestPosts, type SocialPost } from './digestsocial';

export interface MonthStat {
  /** `YYYY-MM` */
  month: string;
  orders: number;
  cancelled: number;
  /** V převažující měně měsíce — koruny se s eury nesčítají */
  revenue: number;
  currency: string;
  items: number;
  /** Kolik různých zákazníků (podle e-mailu) */
  customers: number;
  /** Uzavřený měsíc se už nepřepočítává */
  complete: boolean;
}

export interface SeasonHint {
  /** `YYYY-MM` měsíce, o kterém je řeč */
  month: string;
  label: string;
  /**
   * Jak se té sezóně říká.
   *
   * Index měsíce je z dat, jméno z kalendáře: listopad a prosinec jsou
   * Vánoce, květen až září svatby. Jméno nic nepočítá, jen říká, o čem
   * je řeč — pro e-shop s kravatami a kšandami jsou to dvě různé věci
   * s různým zbožím.
   */
  name: string;
  /** Kolikrát silnější než průměrný měsíc (1,6 = o 60 % víc) */
  index: number;
  /** Do kdy se má začít, ať to má náběh */
  startBy: string;
  /** Za kolik dní sezóna začíná (0 = už běží) */
  inDays: number;
  text: string;
  basis: string;
  /** Co se v ní historicky prodávalo nejvíc */
  products: { code: string; title: string; qty: number }[];
  /** Které příspěvky v tom období fungovaly — podklad pro chystanou kampaň */
  posts: SocialPost[];
}

export interface HistoryView {
  months: MonthStat[];
  /** Kolik měsíců feed vůbec pokrývá — pod rok se sezóny nepočítají */
  coverage: number;
  /** Stejných třicet dní loni; `null`, když data tak daleko nesahají */
  lastYear: { orders: number; revenue: number } | null;
  /** Kolik z dvanácti měsíců bylo slabších než současné okno */
  rank: { better: number; of: number } | null;
  /** Nejbližší sezóna, na kterou se vyplatí chystat */
  season: SeasonHint | null;
  /**
   * Proč sezóna není.
   *
   * Prázdné místo je nejhorší odpověď: nedá se z něj poznat, jestli se
   * nepočítalo, nebo jestli fakt žádná sezóna nepřichází. Tohle se ukáže
   * vždycky, když `season` chybí.
   */
  seasonNote: string;
}

const MONTHS = ['leden', 'únor', 'březen', 'duben', 'květen', 'červen',
  'červenec', 'srpen', 'září', 'říjen', 'listopad', 'prosinec'];

function ensureTable(): void {
  getDb().exec(`
    CREATE TABLE IF NOT EXISTS digest_months (
      month TEXT PRIMARY KEY,
      orders INTEGER NOT NULL DEFAULT 0,
      cancelled INTEGER NOT NULL DEFAULT 0,
      revenue REAL NOT NULL DEFAULT 0,
      currency TEXT NOT NULL DEFAULT '',
      items INTEGER NOT NULL DEFAULT 0,
      customers INTEGER NOT NULL DEFAULT 0,
      complete INTEGER NOT NULL DEFAULT 0,
      computed_at TEXT NOT NULL DEFAULT ''
    );
  `);
}

function monthKey(date: Date): string {
  return `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, '0')}`;
}

function isCancelled(status: string): boolean {
  return /storn|zru[šs]en|vr[áa]cen|cancel|refund/i.test(status ?? '');
}

/**
 * Spočítá jeden měsíc z feedu.
 *
 * Storna se počítají zvlášť a do tržby nejdou — jinak by měsíc s jednou
 * velkou stornovanou objednávkou vypadal jako rekordní.
 */
function computeMonth(month: string): MonthStat {
  const rows = getDb().prepare(
    `SELECT status, currency, total, email, items_json FROM shop_orders
      WHERE substr(created_at, 1, 7) = ?`
  ).all(month) as any[];

  const money = new Map<string, number>();
  const customers = new Set<string>();
  let orders = 0;
  let cancelled = 0;
  let items = 0;

  for (const row of rows) {
    orders++;
    const email = String(row.email ?? '').trim().toLowerCase();
    if (email) customers.add(email);
    if (isCancelled(String(row.status ?? ''))) { cancelled++; continue; }
    const currency = String(row.currency || 'CZK').toUpperCase();
    money.set(currency, (money.get(currency) ?? 0) + Number(row.total || 0));
    try {
      for (const item of JSON.parse(row.items_json || '[]')) items += Number(item?.quantity) || 0;
    } catch { /* rozbitý řádek neshodí měsíc */ }
  }

  const best = [...money.entries()].sort((a, b) => b[1] - a[1])[0];
  return {
    month,
    orders,
    cancelled,
    revenue: Math.round(best?.[1] ?? 0),
    currency: best?.[0] ?? '',
    items,
    customers: customers.size,
    complete: month < monthKey(new Date())
  };
}

/**
 * Měsíční souhrny.
 *
 * Uzavřené měsíce se berou z tabulky, rozdělaný se počítá vždy znovu.
 * Zvlášť se přepočítá i **poslední uzavřený** — feed dobíhá a objednávka
 * z posledního dne měsíce může dorazit až prvního.
 */
export function monthlyStats(count = 13): MonthStat[] {
  ensureTable();
  const d = getDb();

  const first = (d.prepare(
    "SELECT MIN(substr(created_at, 1, 7)) AS m FROM shop_orders WHERE created_at != ''"
  ).get() as any)?.m as string | undefined;
  if (!first) return [];

  const now = new Date();
  const wanted: string[] = [];
  for (let back = count - 1; back >= 0; back--) {
    const when = new Date(now.getFullYear(), now.getMonth() - back, 1);
    const key = monthKey(when);
    if (key >= first) wanted.push(key);
  }

  const cached = new Map<string, MonthStat>();
  for (const row of d.prepare('SELECT * FROM digest_months').all() as any[]) {
    cached.set(row.month, {
      month: row.month, orders: row.orders, cancelled: row.cancelled,
      revenue: row.revenue, currency: row.currency, items: row.items,
      customers: row.customers, complete: !!row.complete
    });
  }

  const write = d.prepare(
    `INSERT INTO digest_months (month, orders, cancelled, revenue, currency, items, customers, complete, computed_at)
     VALUES (?,?,?,?,?,?,?,?,?)
     ON CONFLICT(month) DO UPDATE SET
       orders = excluded.orders, cancelled = excluded.cancelled, revenue = excluded.revenue,
       currency = excluded.currency, items = excluded.items, customers = excluded.customers,
       complete = excluded.complete, computed_at = excluded.computed_at`
  );

  const thisMonth = monthKey(now);
  const lastClosed = monthKey(new Date(now.getFullYear(), now.getMonth() - 1, 1));

  const out: MonthStat[] = [];
  for (const month of wanted) {
    const known = cached.get(month);
    const stale = !known || !known.complete || month === thisMonth || month === lastClosed;
    const stat = stale ? computeMonth(month) : known;
    if (stale) {
      write.run(stat.month, stat.orders, stat.cancelled, stat.revenue, stat.currency,
        stat.items, stat.customers, stat.complete ? 1 : 0, new Date().toISOString());
    }
    out.push(stat);
  }
  return out;
}

/** Objednávky a tržba v libovolném rozsahu dnů — na srovnání s loňskem */
function totalsBetween(fromDay: string, toDay: string, currency: string): { orders: number; revenue: number } {
  const rows = getDb().prepare(
    `SELECT status, currency, total FROM shop_orders
      WHERE substr(created_at, 1, 10) >= ? AND substr(created_at, 1, 10) <= ?`
  ).all(fromDay, toDay) as any[];

  let orders = 0;
  let revenue = 0;
  for (const row of rows) {
    orders++;
    if (isCancelled(String(row.status ?? ''))) continue;
    if (String(row.currency || 'CZK').toUpperCase() !== currency) continue;
    revenue += Number(row.total || 0);
  }
  return { orders, revenue: Math.round(revenue) };
}

function dayKey(date: Date): string {
  const pad = (n: number) => String(n).padStart(2, '0');
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
}

/**
 * Sezóny z vlastních dat.
 *
 * Index měsíce je jeho průměrný denní počet objednávek dělený celoročním
 * průměrem. Prosinec s indexem 1,8 znamená, že se v něm prodávalo skoro
 * dvakrát tolik co obvykle. **Nic se nepředpokládá dopředu**: když e-shop
 * žádnou sezónu nemá, žádná se nenajde.
 *
 * Hlásí se nejbližší měsíc, který ještě nezačal a jehož index je aspoň
 * o čtvrtinu nad průměrem — a k němu datum, do kterého se má začít.
 * Tři týdny předem je odhad postavený na tom, že objednávky na dárky
 * začínají chodit dřív než v samotném měsíci.
 */
function seasonFrom(months: MonthStat[], now: Date): { season: SeasonHint | null; note: string } {
  /*
   * Kolik měsíců stačí. Dvanáct byl původní požadavek a v praxi znamenal,
   * že se sezóna neukázala nikdy — feed tak daleko nesahá. Půl roku stačí
   * na to, aby se dal porovnat nejbližší měsíc s ostatními; jistota je
   * menší, tak se u kratší historie chce výraznější rozdíl a v podkladu
   * je vidět, z kolika měsíců se počítalo.
   */
  const closed = months.filter(one => one.complete);
  if (closed.length < 6) {
    return {
      season: null,
      note: `Na sezónu zatím není dost historie — uzavřených měsíců je ${closed.length}, `
        + 'porovnávat se dá od šesti.'
    };
  }

  const perMonth = new Map<number, { orders: number; days: number }>();
  for (const one of closed) {
    const index = Number(one.month.slice(5, 7)) - 1;
    const [year, month] = one.month.split('-').map(Number);
    const days = new Date(year, month, 0).getDate();
    const found = perMonth.get(index) ?? { orders: 0, days: 0 };
    found.orders += one.orders;
    found.days += days;
    perMonth.set(index, found);
  }

  const daily = new Map<number, number>();
  let sum = 0;
  for (const [index, one] of perMonth) {
    const value = one.orders / Math.max(1, one.days);
    daily.set(index, value);
    sum += value;
  }
  const average = sum / Math.max(1, daily.size);
  if (average <= 0) return { season: null, note: 'Ve feedu nejsou objednávky, ze kterých by šla sezóna poznat.' };
  // Kratší historie snese víc náhody, proto se u ní chce větší rozdíl
  const threshold = closed.length >= 12 ? 1.2 : 1.3;

  /*
   * Nejbližší měsíc, který teprve přijde. Půl roku dopředu: na Vánoce se
   * kampaň chystá v září a „za dva měsíce" je přesně ta zpráva, která se
   * hodí — se čtyřměsíčním výhledem se v létě neukázalo nic.
   */
  const upcoming: { index: number; ratio: number }[] = [];
  for (let ahead = 0; ahead <= 5; ahead++) {
    const when = new Date(now.getFullYear(), now.getMonth() + ahead, 1);
    const index = when.getMonth();
    const value = daily.get(index);
    if (value == null) continue;
    const ratio = value / average;
    upcoming.push({ index, ratio });
    if (ratio < threshold) continue;

    // Už běží? Pak se nemá co chystat, jen ať se ví, v čem se je
    const running = ahead === 0;
    const startBy = new Date(when.getTime() - 21 * 86_400_000);
    const label = `${MONTHS[index]}`;
    const name = seasonName(index);
    const inDays = Math.max(0, Math.round((when.getTime() - now.getTime()) / 86_400_000));
    const stronger = Math.round((ratio - 1) * 100);

    /*
     * Co se v té sezóně prodávalo a co se k ní hodilo napsat. Bez toho je
     * z upozornění jen „prosinec bývá silný" — s tím se nedá nic dělat.
     * Bere se **celá historie**, ne jen loňsko: dva prosince řeknou víc
     * než jeden.
     */
    const months = seasonMonths(index);
    const products = seasonProducts(months);
    const posts = bestPosts({ months, limit: 2 });

    const whenText = running
      ? `Běží ${name} (${label})`
      : inDays > 45
        ? `${name.charAt(0).toUpperCase()}${name.slice(1)} se blíží — začíná zhruba za ${Math.round(inDays / 30)} měsíce`
        : `${name.charAt(0).toUpperCase()}${name.slice(1)} se blíží — začíná zhruba za ${inDays} dní`;

    const hint: SeasonHint = {
      month: monthKey(when),
      label,
      name,
      index: Math.round(ratio * 100) / 100,
      startBy: dayKey(startBy),
      inDays: running ? 0 : inDays,
      text: `${whenText}; ${label} bývá o ${stronger} % silnější než průměrný měsíc`
        + (running ? '.' : ` — propagaci zahájit do ${startBy.getDate()}. ${startBy.getMonth() + 1}.`)
        + (products.length
          ? ` Nejvíc se v ní prodávalo: ${products.slice(0, 3).map(one => one.title).join(', ')}.`
          : ''),
      basis: `průměrně ${value.toFixed(1)} objednávky na den proti celoročním ${average.toFixed(1)}`
        + `, z ${closed.length} měsíců historie`,
      products,
      posts
    };
    return { season: hint, note: '' };
  }

  /*
   * Nic nevybočilo. I to je odpověď — jen se musí říct nahlas a s čísly,
   * ať je poznat, že se počítalo a nic se nenašlo.
   */
  const best = upcoming.sort((a, b) => b.ratio - a.ratio)[0];
  const nearest = upcoming.map(one => MONTHS[one.index]).slice(0, 3).join(', ');
  return {
    season: null,
    note: best
      ? `Nejbližší měsíce (${nearest}) z průměru nevybočují — nejsilnější z nich `
        + `${MONTHS[best.index]} je na ${Math.round(best.ratio * 100)} % celoročního průměru, `
        + `sezóna se hlásí od ${Math.round(threshold * 100)} %.`
      : 'Pro nejbližší měsíce zatím nejsou v historii žádná data k porovnání.'
  };
}

/**
 * Jméno sezóny.
 *
 * Sílu měsíce spočítala data, tohle je jen popiska — ale bez ní je rada
 * „chystej se na listopad" o polovinu míň užitečná než „chystej se na
 * Vánoce". Pro e-shop s kravatami a kšandami jsou svatby a Vánoce dvě různé
 * sezóny s jiným zbožím.
 */
function seasonName(monthIndex: number): string {
  if (monthIndex === 10 || monthIndex === 11) return 'vánoční sezóna';
  if (monthIndex >= 4 && monthIndex <= 8) return 'svatební sezóna';
  return `${MONTHS[monthIndex]}`;
}

/** Které měsíce k sezóně patří — Vánoce jsou listopad i prosinec */
function seasonMonths(monthIndex: number): number[] {
  if (monthIndex === 10 || monthIndex === 11) return [10, 11];
  if (monthIndex >= 4 && monthIndex <= 8) return [4, 5, 6, 7, 8];
  return [monthIndex];
}

/**
 * Co se v sezóně prodávalo.
 *
 * Napříč všemi roky, které feed pokrývá — jeden prosinec může být náhoda,
 * dva už ne. Stornované objednávky se nepočítají.
 */
function seasonProducts(months: number[], limit = 5): { code: string; title: string; qty: number }[] {
  let rows: any[] = [];
  try {
    rows = getDb().prepare(
      'SELECT status, created_at, items_json FROM shop_orders WHERE created_at != \'\''
    ).all() as any[];
  } catch {
    return [];
  }

  const wanted = new Set(months);
  const qty = new Map<string, number>();
  const titles = new Map<string, string>();
  for (const row of rows) {
    const month = Number(String(row.created_at ?? '').slice(5, 7)) - 1;
    if (!wanted.has(month)) continue;
    if (isCancelled(String(row.status ?? ''))) continue;
    let items: any[] = [];
    try { items = JSON.parse(row.items_json || '[]'); } catch { continue; }
    for (const item of items) {
      const code = String(item?.code || item?.title || '').trim();
      if (!code) continue;
      qty.set(code, (qty.get(code) ?? 0) + (Number(item?.quantity) || 0));
      if (!titles.has(code)) titles.set(code, String(item?.title || code));
    }
  }

  return [...qty.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, limit)
    .map(([code, count]) => ({ code, title: titles.get(code) ?? code, qty: count }));
}

/**
 * Zasazení posledních třiceti dní do delší historie.
 *
 * `windowOrders` a `currency` přicházejí z přehledu, aby se totéž nepočítalo
 * dvakrát a aby srovnání sedělo na tutéž měnu.
 */
export function historyView(
  windowOrders: number, currency: string, now = new Date()
): HistoryView {
  /*
   * Dva roky. Třináct měsíců stačí na „stejné okno loni", ale na sezónu ne:
   * dva prosince řeknou víc než jeden a s třinácti měsíci nebylo z čeho brát.
   */
  const months = monthlyStats(25);

  // Stejné okno loni — jediné srovnání, které nemate sezónou
  const from = new Date(now.getTime() - 29 * 86_400_000);
  const lastYearFrom = new Date(from.getFullYear() - 1, from.getMonth(), from.getDate());
  const lastYearTo = new Date(now.getFullYear() - 1, now.getMonth(), now.getDate());
  const oldest = (getDb().prepare(
    "SELECT MIN(substr(created_at, 1, 10)) AS d FROM shop_orders WHERE created_at != ''"
  ).get() as any)?.d as string | undefined;
  const lastYear = oldest && oldest <= dayKey(lastYearFrom)
    ? totalsBetween(dayKey(lastYearFrom), dayKey(lastYearTo), currency)
    : null;

  /*
   * Kolikátý je současný měsíc mezi dvanácti předchozími. Je to hrubé —
   * měsíce nejsou stejně dlouhé — ale odpovídá to na otázku, kterou si
   * člověk klade: „je tohle hodně, nebo málo?"
   */
  const closed = months.filter(one => one.complete).slice(-12);
  const rank = closed.length >= 3
    ? { better: closed.filter(one => one.orders < windowOrders).length, of: closed.length }
    : null;

  const season = seasonFrom(months, now);
  return {
    months,
    coverage: months.length,
    lastYear,
    rank,
    season: season.season,
    seasonNote: season.note
  };
}

File: ios/QuentinoApp/Services/DigestHistory.swift
import Foundation

/**
 Dlouhodobá čísla k AI přehledu.

 Sesterský modul k `src/main/digesthistory.ts`. Třicet dní samo o sobě
 neřekne, jestli je to hodně nebo málo — „sto třináct objednávek" je dobrá
 zpráva v lednu a špatná v prosinci. Drží se proto měsíční souhrny za celou
 historii ve feedu a z nich se počítá zasazení do roku, srovnání s loňskem
 a sezóny.

 Sezóny se **nehádají podle kalendáře**: index měsíce se počítá z vlastních
 dat, takže Vánoce, svatby i letní útlum vyjdou samy, pokud v číslech
 doopravdy jsou. Když e-shop žádnou sezónu nemá, žádná se nenajde.

 Uzavřené měsíce se ukládají do `digest_months` — procházet tisíce
 objednávek při každém otevření je zbytečné, když se už nezmění.
 */
enum DigestHistory {
    private static let monthNames = ["leden", "únor", "březen", "duben", "květen", "červen",
                                     "červenec", "srpen", "září", "říjen", "listopad", "prosinec"]

    private static func ensureTable() {
        _ = try? SQLite.shared.run("""
            CREATE TABLE IF NOT EXISTS digest_months (
              month TEXT PRIMARY KEY,
              orders INTEGER NOT NULL DEFAULT 0,
              cancelled INTEGER NOT NULL DEFAULT 0,
              revenue REAL NOT NULL DEFAULT 0,
              currency TEXT NOT NULL DEFAULT '',
              items INTEGER NOT NULL DEFAULT 0,
              customers INTEGER NOT NULL DEFAULT 0,
              complete INTEGER NOT NULL DEFAULT 0,
              computed_at TEXT NOT NULL DEFAULT ''
            )
            """)
    }

    private static func monthKey(_ date: Date) -> String {
        let parts = Calendar.current.dateComponents([.year, .month], from: date)
        return String(format: "%04d-%02d", parts.year ?? 0, parts.month ?? 0)
    }

    private static func dayKey(_ date: Date) -> String {
        let parts = Calendar.current.dateComponents([.year, .month, .day], from: date)
        return String(format: "%04d-%02d-%02d", parts.year ?? 0, parts.month ?? 0, parts.day ?? 0)
    }

    private static func isCancelled(_ status: String) -> Bool {
        status.range(of: "storn|zru[šs]en|vr[áa]cen|cancel|refund",
                     options: [.regularExpression, .caseInsensitive]) != nil
    }

    /// Spočítá jeden měsíc z feedu; storna do tržby nejdou
    private static func computeMonth(_ month: String) -> [String: Any] {
        let rows = (try? SQLite.shared.query(
            "SELECT status, currency, total, email, items_json FROM shop_orders "
            + "WHERE substr(created_at, 1, 7) = ?", [.text(month)])) ?? []

        var money: [String: Double] = [:]
        var customers = Set<String>()
        var orders = 0
        var cancelled = 0
        var items = 0

        for row in rows {
            orders += 1
            let email = (row["email"] as? String ?? "").trimmingCharacters(in: .whitespaces).lowercased()
            if !email.isEmpty { customers.insert(email) }
            if isCancelled(row["status"] as? String ?? "") { cancelled += 1; continue }
            let currency = (row["currency"] as? String ?? "CZK").uppercased()
            let total = row["total"] as? Double ?? Double(row["total"] as? Int ?? 0)
            money[currency] = (money[currency] ?? 0) + total
            if let text = row["items_json"] as? String, let data = text.data(using: .utf8),
               let list = (try? JSONSerialization.jsonObject(with: data)) as? [[String: Any]] {
                for item in list {
                    items += item["quantity"] as? Int ?? Int(item["quantity"] as? Double ?? 0)
                }
            }
        }

        let best = money.sorted { $0.value > $1.value }.first
        var out: [String: Any] = [:]
        out["month"] = month
        out["orders"] = orders
        out["cancelled"] = cancelled
        out["revenue"] = Int((best?.value ?? 0).rounded())
        out["currency"] = best?.key ?? ""
        out["items"] = items
        out["customers"] = customers.count
        out["complete"] = month < monthKey(Date())
        return out
    }

    /**
     Měsíční souhrny.

     Uzavřené měsíce se berou z tabulky, rozdělaný se počítá vždy znovu —
     a s ním i poslední uzavřený, protože feed dobíhá a objednávka
     z posledního dne může dorazit až prvního.
     */
    static func monthlyStats(_ count: Int = 13, now: Date = Date()) -> [[String: Any]] {
        ensureTable()
        guard let first = (try? SQLite.shared.query(
            "SELECT MIN(substr(created_at, 1, 7)) AS m FROM shop_orders WHERE created_at != ''"
        ))?.first?["m"] as? String, !first.isEmpty else { return [] }

        var wanted: [String] = []
        for back in stride(from: count - 1, through: 0, by: -1) {
            guard let when = Calendar.current.date(byAdding: .month, value: -back, to: now) else { continue }
            let key = monthKey(when)
            if key >= first { wanted.append(key) }
        }

        var cached: [String: [String: Any]] = [:]
        for row in (try? SQLite.shared.query("SELECT * FROM digest_months")) ?? [] {
            if let month = row["month"] as? String { cached[month] = row }
        }

        let thisMonth = monthKey(now)
        let lastClosed = monthKey(Calendar.current.date(byAdding: .month, value: -1, to: now) ?? now)

        var out: [[String: Any]] = []
        for month in wanted {
            let known = cached[month]
            let complete = (known?["complete"] as? Int ?? 0) == 1
            let stale = known == nil || !complete || month == thisMonth || month == lastClosed
            var stat: [String: Any]
            if stale {
                stat = computeMonth(month)
                _ = try? SQLite.shared.run(
                    "INSERT OR REPLACE INTO digest_months "
                    + "(month, orders, cancelled, revenue, currency, items, customers, complete, computed_at) "
                    + "VALUES (?,?,?,?,?,?,?,?,?)",
                    [.text(month),
                     .int(Int64(stat["orders"] as? Int ?? 0)),
                     .int(Int64(stat["cancelled"] as? Int ?? 0)),
                     .double(Double(stat["revenue"] as? Int ?? 0)),
                     .text(stat["currency"] as? String ?? ""),
                     .int(Int64(stat["items"] as? Int ?? 0)),
                     .int(Int64(stat["customers"] as? Int ?? 0)),
                     .int((stat["complete"] as? Bool ?? false) ? 1 : 0),
                     .text(Formats.iso(Date()))]
                )
            } else {
                var known = known ?? [:]
                known["complete"] = complete
                known["revenue"] = Int(known["revenue"] as? Double ?? Double(known["revenue"] as? Int ?? 0))
                stat = known
            }
            out.append(stat)
        }
        return out
    }

    private static func totalsBetween(_ from: String, _ to: String, _ currency: String) -> (orders: Int, revenue: Int) {
        let rows = (try? SQLite.shared.query(
            "SELECT status, currency, total FROM shop_orders "
            + "WHERE substr(created_at, 1, 10) >= ? AND substr(created_at, 1, 10) <= ?",
            [.text(from), .text(to)])) ?? []

        var orders = 0
        var revenue = 0.0
        for row in rows {
            orders += 1
            if isCancelled(row["status"] as? String ?? "") { continue }
            if (row["currency"] as? String ?? "CZK").uppercased() != currency { continue }
            revenue += row["total"] as? Double ?? Double(row["total"] as? Int ?? 0)
        }
        return (orders, Int(revenue.rounded()))
    }

    /**
     Sezóna z vlastních dat.

     Index měsíce je jeho průměrný denní počet objednávek dělený celoročním
     průměrem. Hlásí se nejbližší měsíc s indexem aspoň o čtvrtinu nad
     průměrem — a datum, do kterého se má začít chystat.
     */
    private static func season(_ months: [[String: Any]], _ now: Date) -> [String: Any]? {
        let closed = months.filter { ($0["complete"] as? Bool ?? false) }
        guard closed.count >= 12 else { return nil }

        var orders: [Int: Int] = [:]
        var days: [Int: Int] = [:]
        for one in closed {
            guard let month = one["month"] as? String else { continue }
            let parts = month.split(separator: "-").compactMap { Int($0) }
            guard parts.count == 2 else { continue }
            let index = parts[1] - 1
            var components = DateComponents()
            components.year = parts[0]
            components.month = parts[1]
            let inMonth = Calendar.current.date(from: components).map {
                Calendar.current.range(of: .day, in: .month, for: $0)?.count ?? 30
            } ?? 30
            orders[index] = (orders[index] ?? 0) + (one["orders"] as? Int ?? 0)
            days[index] = (days[index] ?? 0) + inMonth
        }

        var daily: [Int: Double] = [:]
        var sum = 0.0
        for (index, count) in orders {
            let value = Double(count) / Double(max(1, days[index] ?? 1))
            daily[index] = value
            sum += value
        }
        guard !daily.isEmpty else { return nil }
        let average = sum / Double(daily.count)
        guard average > 0 else { return nil }

        for ahead in 0...3 {
            guard let when = Calendar.current.date(byAdding: .month, value: ahead, to: now) else { continue }
            let index = Calendar.current.component(.month, from: when) - 1
            guard let value = daily[index] else { continue }
            let ratio = value / average
            if ratio < 1.25 { continue }

            let startBy = when.addingTimeInterval(-21 * 86_400)
            let label = monthNames[max(0, min(11, index))]
            let percent = Int(((ratio - 1) * 100).rounded())
            let startDay = Calendar.current.component(.day, from: startBy)
            let startMonth = Calendar.current.component(.month, from: startBy)

            var out: [String: Any] = [:]
            out["month"] = monthKey(when)
            out["label"] = label
            out["index"] = (ratio * 100).rounded() / 100
            out["startBy"] = dayKey(startBy)
            out["text"] = ahead == 0
                ? "Běží \(label) — bývá o \(percent) % silnější než průměrný měsíc."
                : "\(label.prefix(1).uppercased())\(label.dropFirst()) bývá o \(percent) % silnější "
                  + "než průměrný měsíc — chystat se má do \(startDay). \(startMonth)."
            out["basis"] = String(format: "průměrně %.1f objednávky na den proti celoročním %.1f, z %d měsíců historie",
                                  value, average, closed.count)
            return out
        }
        return nil
    }

    /// Zasazení posledních třiceti dní do delší historie
    static func view(windowOrders: Int, currency: String, now: Date = Date()) -> [String: Any] {
        let months = monthlyStats(13, now: now)

        let from = now.addingTimeInterval(-29 * 86_400)
        var lastYear: Any = NSNull()
        if let yearFrom = Calendar.current.date(byAdding: .year, value: -1, to: from),
           let yearTo = Calendar.current.date(byAdding: .year, value: -1, to: now),
           let oldest = (try? SQLite.shared.query(
               "SELECT MIN(substr(created_at, 1, 10)) AS d FROM shop_orders WHERE created_at != ''"
           ))?.first?["d"] as? String,
           oldest <= dayKey(yearFrom) {
            let totals = totalsBetween(dayKey(yearFrom), dayKey(yearTo), currency)
            var one: [String: Any] = [:]
            one["orders"] = totals.orders
            one["revenue"] = totals.revenue
            lastYear = one
        }

        let closed = months.filter { ($0["complete"] as? Bool ?? false) }.suffix(12)
        var rank: Any = NSNull()
        if closed.count >= 3 {
            let better = closed.filter { ($0["orders"] as? Int ?? 0) < windowOrders }.count
            var one: [String: Any] = [:]
            one["better"] = better
            one["of"] = closed.count
            rank = one
        }

        var out: [String: Any] = [:]
        out["months"] = months
        out["coverage"] = months.count
        out["lastYear"] = lastYear
        out["rank"] = rank
        out["season"] = season(months, now) ?? NSNull()
        return out
    }
}
